src/mlpack/tests/main_tests/nmf_test.cpp

```diff
@@ -0,0 +1,287 @@
+  SetInputParam("input", std::move(v));
+  SetInputParam("rank", r);
+
+  // Perform NMF.
+  mlpackMain();
+
+  // Get resulting matrices.
+  const mat& w = CLI::GetParam<mat>("w");
+  const mat& h = CLI::GetParam<mat>("h");
+
+  // Check the shapes of W and H.
+  BOOST_REQUIRE_EQUAL(w.n_rows, 10);
+  BOOST_REQUIRE_EQUAL(w.n_cols, 5);
+  BOOST_REQUIRE_EQUAL(h.n_rows, 5);
+  BOOST_REQUIRE_EQUAL(h.n_cols, 10);
+}
+
+/**
+ * Ensure the rank is positive.
+ */
+BOOST_AUTO_TEST_CASE(NMFRankBoundTest)
+{
+  mat v = randu<mat>(10, 10);
+  int r;
+
+  // Rank should not be negative.
+  r = -1;
+  SetInputParam("input", std::move(v));
+  SetInputParam("rank", r);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+
+  // Rank should not be 0.
+  r = 0;
+  SetInputParam("rank", r);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Ensure the max_iterations is non-negative.
+ */
+BOOST_AUTO_TEST_CASE(NMFMaxIterartionBoundTest)
+{
+  mat v = randu<mat>(10, 10);
+  int r = 5;
+
+  // max_iterations should be non-negative.
+  SetInputParam("max_iterations", int(-1));
+  SetInputParam("input", std::move(v));
+  SetInputParam("rank", r);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Ensure the update rule is one of 
+ * {"multdist", "multdiv", "als"}.
+ */
+BOOST_AUTO_TEST_CASE(NMFUpdateRuleTest)
+{
+  mat v = randu<mat>(10, 10);
+  int r = 5;
+
+  // Invalid update rule.
+  SetInputParam("update_rules", std::string("invalid_rule"));
+  SetInputParam("input", std::move(v));
+  SetInputParam("rank", r);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Ensure min_residue is used, by testing that 
+ * min_resude makes a difference to the program.  
+ */
+BOOST_AUTO_TEST_CASE(NMFMinResidueTest)
+{
+  mat v = arma::randu(10, 10);
+  mat initialW = arma::randu(10, 5);
+  mat initialH = arma::randu(5, 10);
+  int r = 5;
+
+  // Set a larger min_residue.
+  SetInputParam("min_residue", double(1));
+  SetInputParam("input", v);
+  SetInputParam("rank", r);
+  SetInputParam("initial_w", initialW);
+  SetInputParam("initial_h", initialH);
+
+  mlpackMain();
+
+  const mat w1 = CLI::GetParam<mat>("w");
+  const mat h1 = CLI::GetParam<mat>("h");
+
+  ResetSettings();
+
+  // Set a smaller min_residue.
+  SetInputParam("min_residue", double(1e-3));
+  SetInputParam("input", v);
+  SetInputParam("rank", r);
+  SetInputParam("initial_w", initialW);
+  SetInputParam("initial_h", initialH);
+
+  mlpackMain();
+
+  const mat w2 = CLI::GetParam<mat>("w");
+  const mat h2 = CLI::GetParam<mat>("h");
+
+  // The resulting matrices should be different.
+  BOOST_REQUIRE(!arma::approx_equal(w1, w2, "absdiff", 1e-5)
```

Comment:
``approx_equal`` was added in armadillo 6.700 but the minimum version we require is 6.500.0; however ``CheckMatrices`` should work just fine.

Reply:
The `CheckMatrices` in test_tools.hpp tests that two matrices are approximately equal. Is there a function which tests two matrices are NOT approximately equal, as we need in this case?

Reply:
Hm, you are right, there is not.  You could do something like this:

```
BOOST_REQUIRE_GT(arma::norm(w1 - w2), 1e-5)
```

(I haven't checked that but I think it would work.)

Reply:
fixed. Thanks for helping:)